bin/wiki_entity_linking/entity_linker_evaluation.py

```diff
@@ -0,0 +1,247 @@
+        _calculate_precision(true_positives, false_positives),
+        _calculate_recall(true_positives, false_negatives),
+        acc_by_label
+    )
+
+
+def measure_baselines(data, kb):
+    # Measure 3 performance baselines: random selection, prior probabilities, and 'oracle' prediction for upper bound
+    counts_d = dict()
+
+    random_true_positives = 0
+    random_false_positives = 0
+    random_false_negatives = 0
+    random_correct_d = defaultdict(int)
+    random_incorrect_d = defaultdict(int)
+
+    oracle_true_positives = 0
+    oracle_false_positives = 0
+    oracle_false_negatives = 0
+    oracle_correct_d = defaultdict(int)
+    oracle_incorrect_d = defaultdict(int)
+
+    prior_true_positives = 0
+    prior_false_positives = 0
+    prior_false_negatives = 0
+    prior_correct_d = defaultdict(int)
+    prior_incorrect_d = defaultdict(int)
+
+    docs = [d for d, g in data if len(d) > 0]
+    golds = [g for d, g in data if len(d) > 0]
+
+    for doc, gold in zip(docs, golds):
+        correct_entries_per_article = dict()
+        tagged_entries_per_article = {_offset(ent.start_char, ent.end_char): ent for ent in doc.ents}
+        for entity, kb_dict in gold.links.items():
+            start, end = entity
+            for gold_kb, value in kb_dict.items():
+                # only evaluating on positive examples
+                if value:
+                    offset = _offset(start, end)
+                    correct_entries_per_article[offset] = gold_kb
+                    if offset not in tagged_entries_per_article:
+                        random_false_negatives += 1
+                        oracle_false_negatives += 1
+                        prior_false_negatives += 1
+
+        for ent in doc.ents:
+            label = ent.label_
+            start = ent.start_char
+            end = ent.end_char
+            offset = _offset(start, end)
+            gold_entity = correct_entries_per_article.get(offset, None)
+
+            # the gold annotations are not complete so we can't evaluate missing annotations as 'wrong'
+            if gold_entity is not None:
+                counts_d[label] = counts_d.get(label, 0) + 1
+                candidates = kb.get_candidates(ent.text)
+                oracle_candidate = ""
+                best_candidate = ""
+                random_candidate = ""
+                if candidates:
+                    scores = []
+
+                    for c in candidates:
+                        scores.append(c.prior_prob)
+                        if c.entity_ == gold_entity:
+                            oracle_candidate = c.entity_
+
+                    best_index = scores.index(max(scores))
+                    best_candidate = candidates[best_index].entity_
+                    random_candidate = random.choice(candidates).entity_
+
+                random_correct_d[label] += (gold_entity == random_candidate)
+                random_incorrect_d[label] += (not gold_entity == random_candidate)
+
```

Comment:
This code may be a little shorter than the original `if ... else` structure, but it certainly can't be faster: you're performing the same comparison 4 or 5 times instead of just once...

Reply:
Good point! The way I thought about it the comparison was fast enough to be negligible but I get that this may not be convincing. I'll change it to do only one comparison.

Reply:
I wouldn't have thought speed would be much of a factor in evaluation code? I usually like to keep the evaluation code shorter, because it's sort of a detail.

I'm a bit confused by `not gold_entity == random_candidate` though! Can this just be `gold_entity != random_candidate`?

Reply:
I agree with you @honnibal that speed doesn't really matter here, but in two pretty much equivalent ways of writing the same code I would still prefer the one where the same statement is not evaluated 5 times. Anyway these kind of rewritings are often a matter of taste. I think the last version committed is a nice compromise (and doesn't have the `not ... ==` anymore) ;-)